fossas/spectrometer
# Migrates from using `go -m list all` to `go list -m -json all` format (pass#1)

# Overview

This modifies preferred go modules analysis. 

We use, `go list -m -json all` now, which provides indication if dependency is Direct or Indirect, and if module is Main or Not. This ensures we only label direct dependency as _direct_. 

## Acceptance criteria

- Does not incorrectly label deep dependency as direct dependency

## Testing plan

Start a golang project with:
```
// go.mod
module example.com/foo/bar
go 1.16
require (
	github.com/mitchellh/cli v1.1.2
	gopkg.in/yaml.v3 v3.0.0-20210107192922-496545a6307b
)
```
and
```
// main.go
package main

import (
	"fmt"
	"os"
	"gopkg.in/yaml.v3"
	"github.com/mitchellh/cli"
)

func main() {
	fmt.Println("hello world")
	c := cli.NewCLI("app", "1.0.0")
	fmt.Println(c)

	os.Exit(0)
}
```

1. Run analysis with current fossa version: `fossa analyze -o | jq`
2. Compare against PR's fossa executable by running: `fossa analyze -o | jq`
3. Notice difference in total count of dependency, and total count of direct dependency, between (1) and (2)
4. Print graph from steps (1) and (2) for comparison. 

## Risks

- Since this is preferred tactic, we may have high blast zone, but I don't anticipate substantial risk.

## TBD

- [x] Test against older golang version
- [x] Test against golang submodule type projects
- [x] Test against golang multi package projects (e.g. consul)

## References

Works on https://github.com/fossas/team-analysis/issues/813

## Checklist
